include/motis/gbfs/lru_cache.h

```diff
@@ -0,0 +1,162 @@
+  }
+
+  lru_cache& operator=(lru_cache const& o) {
+    if (this != &o) {
+      auto read_lock = std::shared_lock{o.mutex_};
+      auto write_lock = std::unique_lock{mutex_};
+      max_size_ = o.max_size_;
+      cache_map_ = o.cache_map_;
+      lru_order_ = o.lru_order_;
+      pending_computations_.clear();
+    }
+    return *this;
+  }
+
+  template <typename F>
+  std::shared_ptr<Value> get_or_compute(Key const key, F compute_fn) {
+    // check with shared lock if entry already exists
+    {
+      auto read_lock = std::shared_lock{mutex_};
+      if (auto it = cache_map_.find(key); it != cache_map_.end()) {
+        move_to_front(key);
+        return it->second->value_;
+      }
+    }
+
+    // not found -> acquire exclusive lock to modify the cache
+    auto write_lock = std::unique_lock{mutex_};
+
+    // check again in case another thread inserted it
+    if (auto it = cache_map_.find(key); it != cache_map_.end()) {
+      move_to_front(key);
+      return it->second->value_;
+    }
+
+    // if another thread is already computing it, wait for it
+    if (auto it = pending_computations_.find(key);
+        it != pending_computations_.end()) {
+      auto future = it->second;
+      write_lock.unlock();
+      return future.get();
+    }
+
+    // create pending computation
+    auto promise = std::promise<std::shared_ptr<Value>>{};
+    auto shared_future = promise.get_future().share();
+    pending_computations_[key] = shared_future;
+    write_lock.unlock();
+
+    // compute the value
+    auto value = compute_fn();
+
+    // store the result
+    write_lock.lock();
+
+    if (lru_order_.size() >= max_size_) {
+      // evict least recently used cache entry
+      auto const last_key = lru_order_.back();
+      cache_map_.erase(last_key);
+      lru_order_.pop_back();
+    }
+
+    cache_map_.try_emplace(
+        key, std::make_shared<cache_entry>(cache_entry{key, value}));
+    lru_order_.insert(lru_order_.begin(), key);
+    pending_computations_.erase(key);
+    promise.set_value(value);
+
+    return value;
+  }
+
+  std::shared_ptr<Value> get(Key const key) {
+    auto read_lock = std::shared_lock{mutex_};
+    if (auto it = cache_map_.find(key); it != cache_map_.end()) {
+      return it->second->value_;
+    }
+    return nullptr;
+  }
+
+  bool contains(Key const key) {
+    auto read_lock = std::shared_lock{mutex_};
+    return cache_map_.find(key) != cache_map_.end();
+  }
+
+  template <typename F>
+  void update_if_exists(Key const key, F update_fn) {
+    auto write_lock = std::unique_lock{mutex_};
+    if (auto it = cache_map_.find(key); it != cache_map_.end()) {
+      it->second->value_ = update_fn(it->second->value_);
+      move_to_front(key);
+    }
+  }
+
+  void remove(Key const key) {
+    auto write_lock = std::unique_lock{mutex_};
+    if (auto it = cache_map_.find(key); it != cache_map_.end()) {
+      if (auto const lru_it =
+              std::find(lru_order_.begin(), lru_order_.end(), key);
```

Comment:
```suggestion
              std::find(lru_order_.begin(), lru_order_.end(), key);
```

`utl::find`

But in general looks expensive in case the cache becomes large?

Reply:
`lru_order_` only stores numbers and I expected the cache size to be <= 1000. Finding a number and moving memory around (in `move_to_front`) in <= 8 KB should be fast enough I hope. We could probably limit `gbfs_provider_idx_t` (the entries in this vector) to `uint16_t` instead of `size_t` to make it more compact.

What alternative did you have in mind? For example, I don't think an additional map `key -> lru index` is very helpful because the indices of all the entries change every time `move_to_front` is called (unless only the first entry is accessed), so keeping that map updated would be expensive.